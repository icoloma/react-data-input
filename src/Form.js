import React from "react";
import PropTypes from "prop-types";

// Form is the (optional) container of BoundComponent instances.
export default class Form extends React.Component {
  constructor() {
    super();
    this.onSubmit = this.onSubmit.bind(this);
    this.validationComponents = [];
    this.ref = React.createRef();
  }

  getChildContext() {
    return {
      form: this
    };
  }

  addValidationComponent(c) {
    this.validationComponents.push(c);
  }

  removeValidationComponent(c) {
    const vc = this.validationComponents;
    const index = vc.indexOf(c);
    index !== -1 && vc.splice(index, 1);
  }

  // validate all child components
  // @return a Promise that resolves to true if validation passes for all fields, false otherwise
  validate() {
    const promises = this.validationComponents.map(c => c.validate());
    return Promise.all(promises).then(errorMessages => {
      return errorMessages.every(errorMessage => !errorMessage);
    });
  }

  // if all validations pass, invokes the onSubmit callback
  // @return a Promise to be invoked after the validation is done
  onSubmit(e) {
    e.preventDefault();
    return this.validate().then(results => {
      if (results) {
        this.props.onSubmit(e);
      } else {
        const firstInvalidElement = this.ref.current.querySelector(
          "[aria-invalid=true]"
        );
        firstInvalidElement && firstInvalidElement.focus();
      }
    });
  }

  render() {
    const { state, ...props } = this.props;
    return (
      <form {...props} onSubmit={this.onSubmit} noValidate ref={this.ref}>
        {this.props.children}
      </form>
    );
  }
}

Form.propTypes = {
  // the callback that will be invoked on submit, if validation passes
  onSubmit: PropTypes.func.isRequired,

  // the state that will be injected into nested elements
  state: PropTypes.object
};

Form.childContextTypes = {
  form: PropTypes.object
};
