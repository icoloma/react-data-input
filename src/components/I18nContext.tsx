import React from "react";
import { defaultEntries } from "../core/Messages";

export const I18nContext = React.createContext(defaultEntries);
