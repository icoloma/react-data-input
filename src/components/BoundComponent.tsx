import React, { useContext, ChangeEvent, useEffect, useState } from "react";
import { Converters, Converter } from "../core/Converters";
import { InputProps } from "../core/ValidationActions";
import { FormContext, FormContextContent } from "./Form";

export interface BoundComponentProps extends InputProps {
  /** type of input component to use */
  elementName: "input" | "select" | "textarea";

  /** propertyName. Will also be used as id if none is specified */
  name: string;

  /** Optional: the converter to use. If not set, the default for the input type will be used */
  converter?: Converter<any>;
}

/**
 * An input/select/textarea component that will bind automatically
 * with the corresponding object propety in the containing `Form`
 */
export function BoundComponent({
  elementName,
  converter,
  name,
  id,
  onChange: originalOnChange,
  children,
  ...props
}: BoundComponentProps) {
  const type = props.type;
  if (process.env.NODE_ENV != "production") {
    if (type === "radio" && !props.defaultValue) {
      throw new Error(`Radio button ${name} requires a defaultValue`);
    }
    if (typeof props.value !== "undefined") {
      throw new Error(`${name} should use defaultValue instead of value`);
    }
  }

  converter = converter || Converters[type] || Converters.text;
  const formContext = useContext<FormContextContent>(FormContext);
  const defaultValue =
    type === "checkbox"
      ? undefined
      : type === "radio"
      ? (props.defaultValue as string)
      : converter.objectToHtml(formContext.getValue(name));
  const defaultChecked =
    type === "checkbox"
      ? formContext.getValue(name)
      : type === "radio"
      ? formContext.getValue(name) == defaultValue
      : undefined;

  function onChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    const objectValue =
      type === "checkbox"
        ? element.checked
        : converter.htmlToObject({ value: element.value, ...props });
    formContext.setValue(name, objectValue);
    originalOnChange && originalOnChange(e);
  }

  useEffect(() => {
    formContext.addValidator(name, props);
    // TODO: should we do something to unmount?
  }, []);

  const error = formContext.errors[name];
  const errorMessageId = error ? `${id}_error` : undefined;

  // ARIA attributes to point to the error message
  const adb = props["aria-describedby"];
  const ariaProps = !errorMessageId
    ? {}
    : {
        "aria-invalid": true,
        "aria-describedby": adb ? `${errorMessageId} ${adb}` : errorMessageId,
      };

  return (
    <>
      {React.createElement(
        elementName,
        {
          name,
          id: id,
          onChange,
          defaultChecked,
          defaultValue,
          ...ariaProps,
          ...props,
        },
        children
      )}
      {error ? (
        <div className="input-error" id={errorMessageId} role="alert">
          {error}
        </div>
      ) : undefined}
    </>
  );
}
