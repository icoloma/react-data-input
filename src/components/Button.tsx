import { ButtonHTMLAttributes, useContext } from "react";
import React from "react";
import { FormContext } from "./Form";

/**
 * A submit button that can display an "in progress" css style. Its use is optional.
 * Use: <Button className="mybutton">Save</Button>
 */
export function Button({
  className = "",
  children,
  ...props
}: ButtonHTMLAttributes<HTMLButtonElement>) {
  const { submitting } = useContext(FormContext);
  return (
    <button
      className={`${submitting ? "in-progress " : ""}${className}`}
      {...props}
    >
      {children}
    </button>
  );
}
