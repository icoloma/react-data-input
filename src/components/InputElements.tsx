import React, { useContext, ChangeEvent, useEffect, useState } from "react";
import { Converters, Converter } from "../core/Converters";
import { InputProps } from "../core/ValidationActions";
import { FormContext, FormContextContent } from "./Form";

export interface BoundComponentProps extends InputProps {
  /** type of input component to use */
  elementName: "input" | "select" | "textarea";

  /** propertyName. Will also be used as id if none is specified */
  name: string;

  /** Optional: the converter to use. If not set, the default for the input type will be used */
  converter?: Converter<any>;
}

/**
 * An input/select/textarea component that will bind automatically
 * with the corresponding object propety in the containing `Form`
 */
export function BoundComponent({
  elementName,
  converter,
  name,
  id,
  onChange: originalOnChange,
  children,
  ...props
}: BoundComponentProps) {
  const type = props.type;
  if (process.env.NODE_ENV != "production") {
    if (type === "radio" && !props.defaultValue) {
      throw new Error(`Radio button ${name} requires a defaultValue`);
    }
    if (typeof props.value !== "undefined") {
      throw new Error(`${name} should use defaultValue instead of value`);
    }
  }

  id = id || type === "radio" ? `${name}_${props.defaultValue}` : name;
  converter = converter || Converters[type] || Converters.text;
  const formContext = useContext<FormContextContent>(FormContext);
  const defaultValue =
    type === "checkbox"
      ? undefined
      : type === "radio"
      ? (props.defaultValue as string)
      : converter.objectToHtml(formContext.getValue(name));
  const defaultChecked =
    type === "checkbox"
      ? formContext.getValue(name)
      : type === "radio"
      ? formContext.getValue(name) == defaultValue
      : undefined;

  function onChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    const objectValue =
      type === "checkbox"
        ? element.checked
        : converter.htmlToObject({ value: element.value, ...props });
    formContext.setValue(name, objectValue);
    originalOnChange && originalOnChange(e);
  }

  useEffect(() => {
    formContext.addValidator(name, props);
    // TODO: should we do something to unmount?
  }, []);

  const error = formContext.errors[name];

  // ARIA attributes to point to the error message
  const errorMessageId = error ? `${id}_error` : undefined;
  const adb = props["aria-describedby"];
  const ariaProps = !error
    ? {}
    : {
        "aria-invalid": true,
        "aria-describedby": adb ? `${errorMessageId} ${adb}` : errorMessageId,
      };

  return (
    <>
      {React.createElement(
        elementName,
        {
          name,
          id: id,
          onChange,
          defaultChecked,
          defaultValue,
          ...ariaProps,
          ...props,
        },
        children
      )}
      {error ? (
        <div className="input-error" id={errorMessageId} role="alert">
          {error}
        </div>
      ) : undefined}
    </>
  );
}

type WrappedBoundComponentProps = Omit<BoundComponentProps, "elementName">;

interface ExtendedInputProps extends WrappedBoundComponentProps {
  type: "text" | "url" | "number" | "checkbox" | "date" | "time" | "radio";
}

export function Input(props: ExtendedInputProps) {
  return <BoundComponent elementName="input" {...props} />;
}

export function Select(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="select" {...props} />;
}

export function TextArea(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="textarea" {...props} />;
}
