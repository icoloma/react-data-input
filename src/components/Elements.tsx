import React from "react";
import { BoundComponent, BoundComponentProps } from "./BoundComponent";

type WrappedBoundComponentProps = Omit<BoundComponentProps, "elementName">;

interface ExtendedInputProps extends WrappedBoundComponentProps {
  type: "text" | "url" | "number" | "checkbox" | "date" | "time" | "radio";
}

export function Input(props: ExtendedInputProps) {
  return <BoundComponent elementName="input" {...props} />;
}

export function Select(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="select" {...props} />;
}

export function TextArea(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="textarea" {...props} />;
}
