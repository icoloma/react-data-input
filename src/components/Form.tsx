import React, {
  FormHTMLAttributes,
  ReactEventHandler,
  createContext,
  useState,
  useContext,
  useEffect,
} from "react";
import {
  filterActionsForProps,
  ValidationResult,
} from "../core/ValidationActions";
import { I18nContext } from "./I18nContext";
import { Messages } from "../core/Messages";
import { Validators, Validator } from "../core/Validators";
import { BoundComponentProps } from "./InputElements";

interface FormProps extends FormHTMLAttributes<HTMLFormElement> {
  /** the object being edited  */
  initialValues: any;

  /** the method to invoke when submitting, after passing all validations */
  onSubmit: (values: any) => Promise<any>;
}

/**
 * The list of all validation errors. The key is the propertyName,
 * the value is the error message
 */
export interface ValidationErrors {
  [propertyName: string]: string;
}

/**
 * Context to share with child BoundComponents
 */
export interface FormContextContent {
  /**
   * All error messages, indexed by property name
   */
  errors: ValidationErrors;

  /**
   * True if this form is submitting the value
   */
  submitting: boolean;

  /**
   * Add a Validator to the list of fields validated by this form
   */
  addValidator(props: BoundComponentProps): void;

  /**
   * Triggered by a change in the form field
   */
  setValue(propertyName: string, value: any): void;

  /**
   * Get value associated to a BoundComponent
   */
  getValue(propertyName: string): any;
}
function undefinedFormContext(): never {
  throw new Error("A <Form> context is required");
}
export const FormContext = createContext<FormContextContent>({
  addValidator: undefinedFormContext,
  setValue: undefinedFormContext,
  getValue: undefinedFormContext,
  errors: {},
  submitting: false,
});

/**
 * Form is the required container  of BoundComponent instances.
 */
export function Form({
  initialValues,
  onSubmit,
  children,
  ...props
}: FormProps) {
  // errors to display, sorted by component key
  const [errors, setErrors] = useState<ValidationErrors>({});

  // object with the data introduced in this form
  const [values, setValues] = useState(initialValues);

  // the list of validators to use for this object
  const [validators, setValidators] = useState<Validators>({});

  // the submission status
  const [submitting, setSubmitting] = useState(false);

  // should focus on the first component with error
  const [focusFirstError, setFocusFirstError] = useState(false);

  // validation rules to apply
  const entries = useContext(I18nContext);
  const messages = new Messages(entries);

  const formContextValue: FormContextContent = {
    errors,
    submitting,
    addValidator: function (props) {
      validators[props.name] = new Validator(props);
      setValidators(validators);
    },

    setValue: function (propertyName, value) {
      values[propertyName] = value;
      if (errors[propertyName]) {
        delete errors[propertyName];
        setErrors({ ...errors });
      }
    },

    getValue(propertyName) {
      return values[propertyName];
    },
  };

  async function validate(): Promise<ValidationErrors> {
    const newErrors: ValidationErrors = {};
    await Promise.all(
      Object.entries(validators).map(async ([propertyName, validator]) => {
        const error: ValidationResult = await validator.validate(
          values[propertyName]
        );
        if (!!error) {
          newErrors[propertyName] = messages.get(error, validator.props);
        }
      })
    );
    setErrors(newErrors);
    if (Object.keys(newErrors).length > 0) {
      setFocusFirstError(true);
    }
    return newErrors;
  }

  const onSubmitHandler: ReactEventHandler = async function (e): Promise<any> {
    e.preventDefault();
    const errors = await validate();
    if (!Object.keys(errors).length) {
      setSubmitting(true);
      return onSubmit(values).finally(() => setSubmitting(false));
    } else {
      return Promise.reject(errors);
    }
  };

  // todo: only do this when fully validating, not by setErrors()

  useEffect(() => {
    if (focusFirstError) {
      const firstInvalidElement = document.body.querySelector<HTMLInputElement>(
        "[aria-invalid=true]"
      );
      firstInvalidElement && firstInvalidElement.focus();
      setFocusFirstError(false);
    }
  }, [errors]);

  return (
    <form {...props} onSubmit={onSubmitHandler} noValidate>
      <FormContext.Provider value={formContextValue}>
        {children}
      </FormContext.Provider>
    </form>
  );
}

/**
 * A fieldset that will disable all input controls while submitting
 */
export function FieldSet({ children, ...props }) {
  const formContext = useContext(FormContext);
  return (
    <fieldset disabled={formContext.submitting} {...props}>
      {children}
    </fieldset>
  );
}
