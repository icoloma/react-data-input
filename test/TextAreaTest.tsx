import React from "react";
import { Form } from "../src/components/Form";
import { create } from "react-test-renderer";
import { TextArea } from "../src/components/InputElements";

function noop(): Promise<string> {
  return Promise.resolve("foobar");
}
describe("TextArea", function () {
  it("should render", function () {
    const tree = create(
      <Form
        onSubmit={noop}
        initialValues={{
          text: "foobaz",
        }}
      >
        <TextArea className="foo" name="text" />
      </Form>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
