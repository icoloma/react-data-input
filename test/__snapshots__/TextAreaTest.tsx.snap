// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea should render 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper "
  >
    <textarea
      className="foo"
      id="text"
      name="text"
      onChange={[Function]}
      value="foobaz"
    />
  </div>
</form>
`;
